File: phonebridge/src/shell/java/amirz/shade/phonebridge/shell/ProjectionShellContext.java
package amirz.shade.phonebridge.shell;

import android.content.Context;
import android.content.MutableContextWrapper;

final class ProjectionShellContext extends MutableContextWrapper {
    static final String SHELL_PACKAGE_NAME = "com.android.shell";

    private static final ProjectionShellContext INSTANCE = new ProjectionShellContext();

    static ProjectionShellContext get() {
        return INSTANCE;
    }

    private ProjectionShellContext() {
        super(ProjectionShellWorkarounds.getSystemContext());
    }

    @Override
    public String getPackageName() {
        return SHELL_PACKAGE_NAME;
    }

    @Override
    public String getOpPackageName() {
        return SHELL_PACKAGE_NAME;
    }

    public int getDeviceId() {
        return 0;
    }

    @Override
    public Context getApplicationContext() {
        return this;
    }
}

File: phonebridge/src/main/java/amirz/shade/phonebridge/shell/ProjectionShellContext.kt
package amirz.shade.phonebridge.shell

import android.content.Context
import android.content.MutableContextWrapper

internal class ProjectionShellContext private constructor() : MutableContextWrapper(
    ProjectionShellWorkarounds.systemContext()
) {
    override fun getPackageName(): String = SHELL_PACKAGE_NAME

    override fun getOpPackageName(): String = SHELL_PACKAGE_NAME

    @Suppress("unused")
    fun getDeviceId(): Int = 0

    override fun getApplicationContext(): Context = this

    companion object {
        const val SHELL_PACKAGE_NAME = "com.android.shell"

        val instance: ProjectionShellContext by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            ProjectionShellContext()
        }
    }
}
